### ISSUE TITLE
[new feature] Sign commit with GPG

### ISSUE THREAD

**Issue by gpongelli** [gitextensions/gitextensions#4238]
**Do you want to request a *feature* or report a *bug*?**
feature

**What is the current behavior?**
GE cannot sign commit.

**What is the expected behavior?**
GE can sign commit.
Git has to be at least [1.7.9](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work)

From [Git commit hel page](https://git-scm.com/docs/git-commit):

> -S[\<keyid\>]
> --gpg-sign[=\<keyid\>]
>
>    GPG-sign commits. The keyid argument is optional and defaults to the committer identity; if specified, it must be stuck to the option without a space.

This means that git commit allows to:

1.  Make an unsigned commit (actual behavior)
2. Make a signed commit with a default key ("-S" option)
3. Make a signed commit with a key different from default one ("-S \<keyid\>" option)

List of steps:

- [ ] UI change discussion to handle the three cases above

- [ ] "git commit" command changes

- [ ] "git commit" command tests

**Comment by gpongelli** [comment]
First idea on UI change: change the commit button similar to “fetch/pull” of main dialog.
With automatic “remember last choice”.

What do you think?

**Comment by pmiossec** [comment]
>First idea on UI change: change the commit button similar to “fetch/pull” of main dialog.
>With automatic “remember last choice”.
>What do you think?

We could also add the option in the already existing `options` menu in the commit form that already contains options that are passed to the commit command...

**Comment by gpongelli** [comment]
@pmiossec  Are you talking about the right menu [here](http://git-extensions-documentation.readthedocs.io/en/latest/commit.html), near the commit message field ?

I don’t know if your idea can be user-friendly, if a user want to sign every commit do he have to use the Options menu each time?
How can he switch from the three available state of the commit command?
The main drawback is that those option (no sign, default key, specific key) are mutually exclusive, not like a flag on a menu item... more similar to a radio box choice .

@RussKie have you any idea? Do you see any GUI implication on those proposal?

**Comment by RussKie** [comment]
I haven't given any  thought to it.
All my attention currently is given to 2.51 release. Once it is out I am
sure I will have an opinion :D

**Comment by pmiossec** [comment]
@gpongelli Yes I'm talking about this one. I don't say it's the good way to do it, I just say it's a possibility to think about since the "Sign-off" feature is already there and works the same way. 

I mean:
* By default, "disabled".
* Could be enabled (and take the commiter identity by default)
* there is a textbox to override the default value

**Comment by gpongelli** [comment]
Sorry but I’ve never used “Sign-off” feature, so don’t know how it works and how it was implemented in GE.
It can be a way to sign commit, let’s hear also other proposal ;)
